File: BionicLimbSimulation_unity/Assets/Scripts/UI/PlaybackUI.cs
using System.IO;
using UnityEngine;
using UnityEngine.UIElements;
using BionicLimb.Core;

namespace BionicLimb.UI
{
    /// <summary>
    /// Binds PlaybackPanel.uxml controls to PlaybackController.
    /// Spec §10.
    /// </summary>
    [RequireComponent(typeof(UIDocument))]
    public class PlaybackUI : MonoBehaviour
    {
        [Header("References")]
        public PlaybackController playbackController;

        // ── UXML element references ───────────────────────────────────────────
        private Label         _gestureLabel;
        private Label         _classifierModeLabel;
        private ProgressBar   _progressBar;
        private Button        _playPauseButton;
        private Button        _prevStepButton;
        private Button        _nextStepButton;
        private Button        _importButton;
        private Label         _errorLabel;

        // ══════════════════════════════════════════════════════════════════════
        // Unity lifecycle
        // ══════════════════════════════════════════════════════════════════════

        private void OnEnable()
        {
            var doc  = GetComponent<UIDocument>();
            var root = doc.rootVisualElement;

            _gestureLabel        = root.Q<Label>("GestureLabel");
            _classifierModeLabel = root.Q<Label>("ClassifierModeLabel");
            _progressBar         = root.Q<ProgressBar>("ProgressBar");
            _playPauseButton     = root.Q<Button>("PlayPauseButton");
            _prevStepButton      = root.Q<Button>("PrevStepButton");
            _nextStepButton      = root.Q<Button>("NextStepButton");
            _importButton        = root.Q<Button>("ImportButton");
            _errorLabel          = root.Q<Label>("ErrorLabel");

            _playPauseButton.clicked += OnPlayPauseClicked;
            _prevStepButton.clicked  += OnPrevStepClicked;
            _nextStepButton.clicked  += OnNextStepClicked;
            _importButton.clicked    += OnImportClicked;

            if (playbackController != null)
            {
                playbackController.OnGestureChanged  += OnGestureChanged;
                playbackController.OnProgressChanged += OnProgressChanged;
                playbackController.OnStateChanged    += OnStateChanged;
            }

            RefreshControls(PlaybackController.PlaybackState.Idle);
        }

        private void OnDisable()
        {
            if (playbackController != null)
            {
                playbackController.OnGestureChanged  -= OnGestureChanged;
                playbackController.OnProgressChanged -= OnProgressChanged;
                playbackController.OnStateChanged    -= OnStateChanged;
            }
        }

        // ══════════════════════════════════════════════════════════════════════
        // Button handlers
        // ══════════════════════════════════════════════════════════════════════

        private void OnPlayPauseClicked()
        {
            if (playbackController == null) return;
            if (playbackController.State == PlaybackController.PlaybackState.Playing)
                playbackController.Pause();
            else
                playbackController.Play();
        }

        private void OnPrevStepClicked()  => playbackController?.PreviousStep();
        private void OnNextStepClicked()  => playbackController?.NextStep();

        private void OnImportClicked()
        {
            string path = PickFile();
            if (!string.IsNullOrEmpty(path))
                playbackController?.LoadFile(path);
        }

        // ══════════════════════════════════════════════════════════════════════
        // Event handlers
        // ══════════════════════════════════════════════════════════════════════

        private void OnGestureChanged(Classification.GestureResult result)
        {
            _gestureLabel.text = $"Gesture: {result.GestureName}  ({result.Confidence * 100f:F0}%)";
        }

        private void OnProgressChanged(float progress)
        {
            _progressBar.value = progress;
        }

        private void OnStateChanged(PlaybackController.PlaybackState state)
        {
            RefreshControls(state);
        }

        // ══════════════════════════════════════════════════════════════════════
        // UI helpers
        // ══════════════════════════════════════════════════════════════════════

        private void RefreshControls(PlaybackController.PlaybackState state)
        {
            bool hasFile = state != PlaybackController.PlaybackState.Idle &&
                           state != PlaybackController.PlaybackState.Error;
            bool isPlaying = state == PlaybackController.PlaybackState.Playing;

            _playPauseButton.SetEnabled(hasFile);
            _prevStepButton.SetEnabled(hasFile);
            _nextStepButton.SetEnabled(hasFile);
            _playPauseButton.text = isPlaying ? "⏸" : "▶";

            if (state == PlaybackController.PlaybackState.Error && playbackController != null)
            {
                _errorLabel.text = playbackController.ErrorMessage;
                _errorLabel.style.display = DisplayStyle.Flex;
            }
            else
            {
                _errorLabel.style.display = DisplayStyle.None;
            }

            if (playbackController != null && !string.IsNullOrEmpty(playbackController.ClassifierMode))
                _classifierModeLabel.text = $"Classifier: {playbackController.ClassifierMode}";
        }

        /// <summary>
        /// Returns a file path to load.
        /// In Editor → native OS file picker.
        /// In build → first .myo.csv file found in StreamingAssets/SampleData.
        /// </summary>
        private static string PickFile()
        {
#if UNITY_EDITOR
            return UnityEditor.EditorUtility.OpenFilePanel(
                "Open EMG Recording", Application.streamingAssetsPath, "csv");
#else
            string dir = Path.Combine(Application.streamingAssetsPath, "SampleData");
            if (!Directory.Exists(dir)) return string.Empty;
            string[] files = Directory.GetFiles(dir, "*.myo.csv");
            return files.Length > 0 ? files[0] : string.Empty;
#endif
        }
    }
}

File: bionic-limb-threejs/src/playback/PlaybackUI.ts
import type { PlaybackController } from './PlaybackController.js';
import type { GestureResult } from '../classification/IGestureClassifier.js';

export class PlaybackUI {
  constructor(controller: PlaybackController) {
    const gestureLabel = document.getElementById('gesture-label')!;
    const classifierModeLabel = document.getElementById('classifier-mode-label')!;
    const progressBar = document.getElementById('progress-bar') as HTMLInputElement;
    const btnPlayPause = document.getElementById('btn-play-pause')!;
    const btnNextStep = document.getElementById('btn-next-step')!;
    const btnPrevStep = document.getElementById('btn-prev-step')!;
    const btnImport = document.getElementById('btn-import')!;
    const errorLabel = document.getElementById('error-label')!;

    const fileInput = document.createElement('input');
    fileInput.type = 'file';
    fileInput.accept = '.csv';
    fileInput.style.display = 'none';
    document.body.appendChild(fileInput);

    btnPlayPause.addEventListener('click', () => {
      if (controller.state === 'Playing') controller.pause();
      else controller.play();
    });

    btnNextStep.addEventListener('click', () => controller.nextStep());
    btnPrevStep.addEventListener('click', () => controller.previousStep());
    btnImport.addEventListener('click', () => fileInput.click());

    fileInput.addEventListener('change', () => {
      const file = fileInput.files?.[0];
      if (!file) return;
      const reader = new FileReader();
      reader.onload = (e) => {
        const text = e.target?.result as string;
        try {
          controller.loadFile(text);
          errorLabel.style.display = 'none';
        } catch (err) {
          errorLabel.textContent = err instanceof Error ? err.message : String(err);
          errorLabel.style.display = 'block';
        }
      };
      reader.readAsText(file);
    });

    controller.addEventListener('gestureChanged', (e: Event) => {
      const result = (e as CustomEvent<GestureResult>).detail;
      gestureLabel.textContent = `${result.gestureName} (${(result.confidence * 100).toFixed(0)}%)`;
    });

    controller.addEventListener('progressChanged', (e: Event) => {
      const progress = (e as CustomEvent<number>).detail;
      progressBar.value = String(progress * 100);
    });

    controller.addEventListener('stateChanged', () => {
      btnPlayPause.textContent = controller.state === 'Playing' ? '\u23F8' : '\u25B6';
      classifierModeLabel.textContent = controller.classifierMode;
    });
  }
}
